### PR TITLE
Adds `@odata.type` property and makes this a required property in schemas that have discriminator objects

### PR DESCRIPTION
**Pull request by irvinesunday**
Fixes https://github.com/microsoft/OpenAPI.NET.OData/issues/243

This PR does the following:

Given an example of this schema, which is a base type, which exposes the discriminator mappings as shown:

MicrosoftTeams-image (2)

The following is accomplished:

1. Adds the @odata.type property in the properties' object
2. Marks the @odata.type property as required
3. In the @odata.type property, sets the default value to '#microsoft.graph.entity
4. Updates tests to validate the above.

After the above, the schema should appear as below:

image

### CODE REVIEW

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs [hunk 1]
```diff
@@ -435,6 +435,7 @@ internal static OpenApiSchema CreateSchemaTypeSchema(this ODataContext context,
             {
                 // The discriminator object is added to structured types which have derived types.
                 OpenApiDiscriminator discriminator = null;
+                string discriminatorPropertyName = "@odata.type";
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs:438]: move to a const outside of the method.

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs [hunk 2]
```diff
@@ -473,6 +474,16 @@ internal static OpenApiSchema CreateSchemaTypeSchema(this ODataContext context,
                     AnyOf = null
                 };
 
+                if (schema.Discriminator != null)
+                {
+                    schema.Properties.TryAdd(discriminatorPropertyName, new OpenApiSchema()
+                    {
+                        Type = "string"                        
+                    });
+                    schema.Properties[discriminatorPropertyName].Default = new OpenApiString("#" + structuredType.FullTypeName());
```
Comment (baywet) [inline src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs:483]:
```suggestion
                        Type = "string",
                       .Default = new OpenApiString("#" + structuredType.FullTypeName()),
                    });
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs (hunk 2 above)
Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs:483]: No. I intentionally set the `Default` outside the `TryAdd()` method so as to capture for all cases, including those where the `discriminatorPropertyName` -> `@odata.type` property already exists in the schema.

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs:483]: in what scenario could the property could have already been added?

Comment (irvinesunday) [reply src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs]: In your [comment](https://github.com/microsoft/OpenAPI.NET.OData/issues/243#issuecomment-1168726146) in the issue you mentioned that some base types are missing this property and that we should walk the inheritance tree to make sure the property is present and add it if not.

Comment (baywet) [reply src/Microsoft.OpenApi.OData.Reader/Generator/OpenApiSchemaGenerator.cs]: Ah my bad, I should probably have added another comment with the latest requirement to clear any confusion.
The current implementation adds the property to all entity or complex types (which is what we now want). TryAdd would never have checked the presence of the property in a base schema, for that we'd have had to walk the allOf properties recursively. But since we don't want to add the property only on the base type anymore this requirement is moot.
I can't foresee a case where the odata type property would already be here besides it being described in the CSDL, and using convention reserved property names would be horribly wrong, and the conversion should fail.

Bottom line, move the default above, and if the tryadd returns false, throw an invalid operation exception saying the property is already present, which is not expected, and the CSDL should be verified.

I hope this makes things clearer, don't hesitate if you have follow up questions.

**Review by baywet — APPROVED** [review]
